plc/int6kboot.1
.TH int6kboot 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
int6kboot - Qualcomm Atheros Powerline Device Boot Utility I

.SH SYNOPSIS
.BR int6kboot
.RI [ options ] 
.BR - N 
.IR file 
.BR - P 
.IR file
.RI [ device ]
.RI [ device ]
[...]

.SH DESCRIPTION
Download and start runtime firmware on any local Atheros powerline device that is in bootloader mode.
A device remains in bootloader mode when it cannot load firmware from flash memory for some reason.
This can happen because there is no flash memory or because the flash memory is blank or corrupted.
This program can start a device or start then flash a device.

.PP
The bootloader only recognizes a subset of Atheros message types.
This program uses that subset to write runtime firmware and parameters into SDRAM then start firmware execution at a known address.
Optionally, this program can then proceed to download the same firmware and parameters, again, and instruct the runtime firmware to write them to flash memory.

.PP
Starting a device is a three phase operation that can only be performed when the bootloader is running.
The first phase writes an SDRAM configuration program into TCM and executes it.
The second phase writes runtime firmware and parameters as images into SDRAM.
The third phase starts firmware execution.
Once the runtime firmware starts, the device will be operational.

.PP
Flashing a device is a two phase operation that can only be performed when the firmware is running.
The first phase downloads runtime firmare and parameters into scratch buffers under firmware control.
The second phase instructs the firmware to write the scratch buffers to flash memory and restart the device.
After flashing, the device will be self-sustaining.

.PP
Since the device address is \fB00:B0:52:00:00:01\fR on boot, the host must send all messages to that address.
This can cause problems if multiple devices are connected to the host and in Bootloader mode.
The devices will have the same firmware and MAC address because their PIBs will be identical.
One should avoid this situation.

.PP
Is probably alright to have other device connected to the host that are running firmware because they will still see but not recognize the messages used in this operation.
If only one device is connected to the host, then is save to flash the device using address \fB00:B0:52:00:00:01\fR.
Otherwise, the host should address messages to the actual device address so that other devices do not attempt to write to their own flash memory at the same time.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBplc\fR man page for an overview and installation instructions.

.SH OPTIONS

.TP
.RB - e
Redirects stderr messages to stdout.
By convention status and error messages are printed on stderr while primary program output is printed on stdout.
This option prints all output on stdout in cases where this is desired.

.TP
.RB - F [ F ]
Write previously downloaded MAC and PIB to NVRAM using VS_MOD_NVM.
Adding a second \fBF\fR here or another -\fBF\fR anywhere on the command line will \fBforce-flash\fR a blank or corrupted NVRAM.
Firmware loaded from NVRAM will treat force-flash as an error.
This option can be used to create factory settings but cannot be used to change them once created.
Subsequent use creates and updates operational settings that can be erased using a factory reset.

.TP
-\fB\i \fIinterface\fR
Select the host Ethernet interface.
All requests are sent via this host interface and only reponses received via this host interface are recognized.
The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface.
This option then takes precedence over either default.

.TP 
-\fBN \fIfilename\fR
Read the named NVM file and write it to the device.
This option and argument are required.
Any file having an invalid NVM file format will be rejected.
No assumptions are made based on filename and no filename conventions are enforced.

.TP
-\fBP \fIfilename\fR
Read the named PIB file and write it to the device.
This option and argument are required.
Any file having an invalid PIB file format will be rejected.
No assumptions are made based on filename and no filename conventions are enforced.

.TP
.RB - q
Suppresses status messages on stderr.

.TP
-\fBt \fImilliseconds\fR
Read timeout in milliseconds.
Values range from 0 through UINT_MAX.
This is the maximum time allowed for a response.
The default is shown in brackets on the program menu.

.TP
.RB - v
Prints additional information on stdout.
In particular, this option dumps outgoing Ethernet packets on stdout.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over other options on the command line.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over other options on the command line.
Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.

.SH ARGUMENTS

.TP
.IR device
The Ethernet hardware address of some powerline device.
More than one address may be specified on the command line.
If more than one address is specified then operations are performed on each device in turn.
The default address is \fBlocal\fR.
as explained in the \fBDEVICES\fR section.

.SH DEVICES
Powerline device addresses are 12 hexadecimal digits in upper, lower or mixed case.
Individual octets may be separated by colons, for clarity, but colons are not required.
For example, "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are valid and equivalent.

.PP
A \fBlocal\fR device is any Atheros Powerline Device connected directly to a host Ethernet interface.
A \fBremote\fR device is any Atheros Powerline Device at the far end of a powerline connection.
A \fBforeign\fR device is any powerline device not manufactured by Atheros.

.PP
Common device addresses have symbolic names that can be used in place of the actual address value.
The following symbolic addresses are recognized by this program and most other toolkit programs.

.TP
.BR all
Equivalent to "broadcast", described next.

.TP
.BR broadcast
A synonym for the Ethernet broadcast address, \fBFF:FF:FF:FF:FF:FF\fR.
All devices, whether local, remote or foreign will respond to this address.

.TP
.BR local
A synonym for the Qualcomm Atheros Local Management Address (LMA), \fB00:B0:52:00:00:01\fR.
All local Atheros devices will respond to this address but remote and foreign devices will not.

.SH REFERENCES
See the Qualcomm Atheros HomePlug AV Firmware Technical Reference Manual for technical information.
See the Qualcomm Atheros Powerline Toolkit Online Documetation for practical information and examples.

.SH EXAMPLES
The -\fBN\fR and -\fBP\fR options are always required.
Their order is not important.
Tne MAC address is not needed because devices default to address 00:B0:52:00:00:01 (the program default address) in bootloader mode.
Once this operation completes, firmware is running in SDRAM but the NVRAM, if present, has not been programmed.
This operation is common for flash-less devices.
If we reset the device at this point then the device will return to bootloader mode.

.PP
   int6kboot -N test.nvm -P test.pib
   # eth0 00:B0:52:00:00:01 Write SDRAM Configuration
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Write to device memory
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Write to device memory
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Start Firmware
   # eth0 00:B0:52:00:00:01 Started.

.PP
The next example does the same thing but performs the extra steps needed to program NVRAM.
It first downloads the firmware image from file \fBtest.nvm\fR and the parameter block image from file \fBtest.pib\fR using VS_WR_MEM then starts firmware execution using VS_ST_MAC.
Once runtime firmware has started, it downloads the same firmware image and parameter block image (again) using VS_WR_MOD and commits them to NVRAM using VS_MOD_NVM.

.PP
   int6kboot -N test.nvm -P test.pib -F
   # eth0 00:B0:52:00:00:01 Write SDRAM Configuration
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Write to device memory
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Write to device memory
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Start Firmware
   # eth0 00:B0:52:00:00:01 Started.
   # eth0 00:B0:52:00:00:01 Write MAC as module
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Write PIB as module
   # eth0 00:B0:52:00:00:01 Written.
   # eth0 00:B0:52:00:00:01 Flash Device
   # eth0 00:B0:52:00:00:01 Flashed.

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information may not be available.
Qualcomm Atheros reserves the right to modify message structure and content in future firmware releases without any obligation to notify or compensate users of this program.

.SH SEE ALSO
.BR plc ( 1 ), 
.BR ampboot ( 1 ), 
.BR chknvm ( 1 ), 
.BR chkpib ( 1 ), 
.BR int6k ( 1 ), 
.BR modpib ( 1 ) 

.SH CREDITS
 Charles Maier



CHANGES
open-plc-utils-0.0.6
---------------

1. Programs mac2pw and mac2pwd implement a new method of generating device passwords. The old method is supported but is not the default. 

open-plc-utils-0.0.5
---------------

open-plc-utils-0.0.4
---------------

open-plc-utils-0.0.3
---------------

1. Added header VisualStudioNET/include/inttype.h to define print specifier constant PRId64 for Windows systems.

2. Replaced conditional compiler code around print specifiers "lld" and "I64d" and replaced single specifier PRId64 as defined by the system.

3. Renamed man pages ending in .7 with .1 to reflect that fact that they will now be installed man1 instead of man7.

4. Replaced const HOMEPLUG_MTYPE with constant ETH_P_HPAV throughout.

5. Correct endianess of FIELD_ETH_TYPE in module coqos_add.c.

6. Updated chipset codes and implemented table-driven chipset detection algorithm.

7. Moved TLVNode structure definition from nvm/nvm.h to tools/tlv.h to prepare for exanded use of TLV structures in future MMEs.


open-plc-utils-0.0.2
---------------

1. Removed references to private vendor-specific MMEs

2. Removed unused archive targets in primary Makefile.

3. Added .gitignore to subdirectories where programs and other objects are compiled.

4. Added conditional __UCLIBC__ compile statement to ARCPPrint.c to account for variable argument definition error during compile; 

5. Added 'ignore' target to cascading Makefiles to re-write local .gitignore files.

6. Corrected email address for Florian Fainelli in module ARPCPrint.

7. Added new HomePlug AV SLAC Protocol MMEs definitions to homeplug.h.

8. Modified makefile 'ignore' target to use tr instead of sed because some sed implementations do not accept '\n'.



plc/int6kboot.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or
 *   without modification, are permitted (subject to the limitations
 *   in the disclaimer below) provided that the following conditions
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of
 *     its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   int6kboot.c - Qualcomm Atheros INT6x00 Powerline Device Bootstrapper
 *
 *
 *   Contributor(s):
 *      Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/


/*====================================================================*"
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <fcntl.h>
#include <ctype.h>
#include <sys/stat.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/symbol.h"
#include "../tools/number.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../plc/plc.h"
#include "../ram/nvram.h"
#include "../ram/sdram.h"
#include "../nvm/nvm.h"
#include "../pib/pib.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../plc/chipset.c"
#include "../plc/Confirm.c"
#include "../plc/Devices.c"
#include "../plc/Display.c"
#include "../plc/Failure.c"
#include "../plc/FlashNVM.c"
#include "../plc/ReadMME.c"
#include "../plc/Request.c"
#include "../plc/SendMME.c"
#include "../plc/WriteMEM.c"
#include "../plc/WriteNVM.c"
#include "../plc/WritePIB.c"
#include "../plc/WaitForReset.c"
#include "../plc/WaitForStart.c"
#include "../plc/WaitForRestart.c"
#include "../plc/InitDevice1.c"
#include "../plc/BootDevice1.c"
#include "../plc/BootFirmware1.c"
#include "../plc/BootParameters1.c"
#include "../plc/WriteFirmware1.c"
#include "../plc/WriteExecuteFirmware1.c"
#include "../plc/WriteExecutePIB.c"
#include "../plc/StartFirmware1.c"
#include "../plc/FlashDevice1.c"
#endif

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/checkfilename.c"
#include "../tools/hexdecode.c"
#include "../tools/hexstring.c"
#include "../tools/todigit.c"
#include "../tools/hexdump.c"
#include "../tools/checksum32.c"
#include "../tools/fdchecksum32.c"
#include "../tools/error.c"
#include "../tools/strfbits.c"
#include "../tools/typename.c"
#endif

#ifndef MAKEFILE

// #include "../ram/sdramfile.c"// #include "../ram/sdrampeek.c"

#endif

#ifndef MAKEFILE
#include "../nvm/nvmfile1.c"
#endif

#ifndef MAKEFILE
#include "../pib/pibfile1.c"
#include "../pib/pibpeek1.c"
#endif

#ifndef MAKEFILE
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#include "../ether/channel.c"
#endif

#ifndef MAKEFILE
#include "../mme/MMECode.c"
#include "../mme/EthernetHeader.c"
#include "../mme/QualcommHeader.c"
#include "../mme/UnwantedMessage.c"
#endif

#ifndef MAKEFILE
#include "../key/keys.c"
#endif

/*====================================================================*
 *
 *   int main (int argc, char const * argv[]);
 *
 *   parse command line, populate plc structure and perform selected
 *   operations; show help summary when asked; see getoptv and putoptv
 *   to understand command line parsing and help summary display; see
 *   plc.h for the definition of struct plc;
 *
 *   the default interface is eth1 because most people use eth0 as
 *   their principle network connection; you can specify another
 *   interface with -i or define environment string PLC to make
 *   that the default interface and save typing;
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"ei:FN:p:P:qt:vx",
		"-N file -P file [device] [device] [...]",
		"Qualcomm Atheros INT6x00 Powerline Device Bootstrapper",
		"e\tredirect stderr to stdout",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"F[F]\tflash [force] NVRAM after firmware start using VS_MOD_NVM",
		"N f\twrite NVM file to device using VS_WR_MEM",
		"P f\twrite PIB file to device using VS_WR_MEM",
		"q\tquiet mode",
		"t n\tread timeout is (n) milliseconds [" LITERAL (CHANNEL_TIMEOUT) "]",
		"v\tverbose mode",
		"x\texit on error",
		(char const *) (0)
	};

#include "../plc/plc.c"

	char firmware [PLC_VERSION_STRING];
	signed c;
	if (getenv (PLCDEVICE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (PLCDEVICE));

#else

		channel.ifname = strdup (getenv (PLCDEVICE));

#endif

	}
	optind = 1;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'F':
			_setbits (plc.module, (VS_MODULE_MAC | VS_MODULE_PIB));
			if (_anyset (plc.flags, PLC_FLASH_DEVICE))
			{
				_setbits (plc.module, VS_MODULE_FORCE);
			}
			_setbits (plc.flags, PLC_FLASH_DEVICE);
			break;
		case 'N':
			if (!checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.NVM.file = open (plc.NVM.name = optarg, O_BINARY|O_RDONLY)) == -1)
			{
				error (1, errno, "%s", plc.NVM.name);
			}
			if (nvmfile1 (&plc.NVM))
			{
				error (1, errno, "Bad firmware file: %s", plc.NVM.name);
			}
			_setbits (plc.flags, PLC_WRITE_MAC);
			break;
		case 'P':
			if (!checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.PIB.file = open (plc.PIB.name = optarg, O_BINARY|O_RDONLY)) == -1)
			{
				error (1, errno, "%s", plc.PIB.name);
			}
			if (pibfile1 (&plc.PIB))
			{
				error (1, errno, "Bad parameter file: %s", plc.PIB.name);
			}
			_setbits (plc.flags, PLC_WRITE_PIB);
			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			_setbits (plc.flags, PLC_SILENCE);
			break;
		case 't':
			channel.timeout = (signed)(uintspec (optarg, 0, UINT_MAX));
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (plc.flags, PLC_VERBOSE);
			break;
		case 'x':
			_setbits (plc.flags, PLC_BAILOUT);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc)
	{
		error (1, ENOTSUP, ERROR_TOOMANY);
	}
	if (plc.PIB.file == -1)
	{
		error (1, ECANCELED, "No PIB file specified");
	}
	if (plc.NVM.file == -1)
	{
		error (1, ECANCELED, "No NVM file specified");
	}
	openchannel (&channel);
	if (!(plc.message = malloc (sizeof (* plc.message))))
	{
		error (1, errno, PLC_NOMEMORY);
	}
	if (WaitForStart (&plc, firmware, sizeof (firmware)))
	{
		Failure (&plc, PLC_NODETECT);
		exit (1);
	}
	if (strcmp (firmware, "BootLoader"))
	{
		Failure (&plc, "Bootloader must be running");
		exit (1);
	}
	if (plc.hardwareID < CHIPSET_INT6300)
	{
		Failure (&plc, "Device must be %s or later. Try program int6kf instead.", chipsetname (CHIPSET_INT6300));
		exit (1);
	}
	if (plc.hardwareID >= CHIPSET_AR7400)
	{
		Failure (&plc, "Program does not support %s or later; try program ampboot instead.", chipsetname (CHIPSET_AR7400));
		exit (1);
	}
	if (!InitDevice1 (&plc))
	{
		if (!BootDevice1 (&plc))
		{
			if (_anyset (plc.flags, PLC_FLASH_DEVICE))
			{
				FlashDevice1 (&plc);
			}
		}
	}
	free (plc.message);
	closechannel (&channel);
	exit (0);
}

